### ISSUE TITLE
Not installable on FreeBSD

### ISSUE THREAD

**Issue by ararslan** [JuliaWeb/MbedTLS.jl#132]
FreeBSD currently isn't supported in BinaryBuilder, so as of 0.5.7 MbedTLS is not installable.

**Comment by tkelman** [comment]
is it not falling back to a source build or using the julia-included copies?

**Comment by ararslan** [comment]
No, it emits an error from BinaryProvider saying that it isn't a supported platform.

**Comment by tkelman** [comment]
When, during import or the `any(!satisfied(p; verbose=verbose) for p in products)` call? Either way that's pretty broken.

**Comment by ararslan** [comment]
The latter.
```
julia> Pkg.add("MbedTLS")
[ Info: Cloning cache of BinaryProvider from https://github.com/JuliaPackaging/BinaryProvider.jl.git
[ Info: Cloning cache of MbedTLS from https://github.com/JuliaWeb/MbedTLS.jl.git
[ Info: Installing BinaryProvider v0.2.5
[ Info: Installing MbedTLS v0.5.8
[ Info: Building MbedTLS
┌ Error: ------------------------------------------------------------
│ # Build failed for MbedTLS
│   exception =
│    LoadError: ArgumentError: Platform `x86_64-unknown-freebsd11.1` is not an officially supported platform
│    Stacktrace:
│     [1] platform_key(::String) at /home/alex/.julia/v0.7/BinaryProvider/src/PlatformNames.jl:167
│     [2] platform_key at /home/alex/.julia/v0.7/BinaryProvider/src/PlatformNames.jl:141 [inlined]
│     [3] (::getfield(BinaryProvider, Symbol("#kw##satisfied")))(::NamedTuple{(:verbose,),Tuple{Bool}}, ::typeof(BinaryProvider.satisfied), ::BinaryProvider.LibraryProduct) at ./<missing>:0
│     [4] (::getfield(Main.__anon__, Symbol("##5#9")))(::BinaryProvider.LibraryProduct) at ./<missing>:0
│     [5] next at ./generator.jl:47 [inlined]
│     [6] _any(::typeof(identity), ::Base.Generator{Array{BinaryProvider.Product,1},getfield(Main.__anon__, Symbol("##5#9"))}, ::Colon) at ./reduce.jl:675
│     [7] any at ./reduce.jl:671 [inlined]
│     [8] any(::Base.Generator{Array{BinaryProvider.Product,1},getfield(Main.__anon__, Symbol("##5#9"))}) at ./reduce.jl:602
│     [9] top-level scope
│     [10] include at ./boot.jl:306 [inlined]
│     [11] include_relative(::Module, ::String) at ./loading.jl:1072
│     [12] include at ./sysimg.jl:29 [inlined]
│     [13] include(::String) at ./loading.jl:1106
│     [14] top-level scope
│     [15] eval at ./boot.jl:309 [inlined]
│     [16] eval at ./sysimg.jl:74 [inlined]
│     [17] evalfile(::String, ::Array{String,1}) at ./loading.jl:1101 (repeats 2 times)
│     [18] #6 at ./none:13 [inlined]
│     [19] cd(::getfield(Main, Symbol("##6#8")){String}, ::String) at ./file.jl:72
│     [20] (::getfield(Main, Symbol("##5#7")))(::IOStream) at ./none:12
│     [21] #open#321(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::getfield(Main, Symbol("##5#7")), ::String, ::Vararg{String,N} where N) at ./iostream.jl:369
│     [22] open(::Function, ::String, ::String) at ./iostream.jl:367
│     [23] top-level scope
│     [24] eval at ./boot.jl:309 [inlined]
│     [25] eval(::Module, ::Expr) at ./sysimg.jl:74
│     [26] exec_options(::Base.JLOptions) at ./client.jl:304
│     [27] _start() at ./client.jl:455
│    in expression starting at /usr/home/alex/.julia/v0.7/MbedTLS/deps/build.jl:36
└ @ Main none:16
```

**Comment by samoconnor** [comment]
This seems like a good example of why the `Pkg` system needs to have per-platform metadata.
e.g. If MbedTLS 0.5.7 had "validated platform" tags for `"*-darwin*"`, `"*-mingw32*"`, `"*-linux*"` (but not for `"*-freebsd*"`) then `Pkg.add` would know to use an older version on freebsd.

It seems that as the number of supported platforms grows it will be increasingly common for a package maintainer to be unable to test every platform (embedded systems, systems not supported by Travis etc). In this case it makes sense to push out a release with "validated platform" tags for the platforms that are covered by CI, and let the community add tags for other platforms one by one as they are tested off line.

**Comment by tkelman** [comment]
You can already list an unsatisfiable julia upper bound for a specific platform in the REQUIRE file. Those platform-specific tags are totally ignored by Pkg3 last I checked, so you'll need to chime in on that issue if you care about it getting implemented there.

**Comment by samoconnor** [comment]
The problem is that the REQUIRE file is an immutable part of a release. At the time of making a release tag it is not generally possible to know what platforms will end up being successfully validated for that release.

**Comment by tkelman** [comment]
That's a false impression you get by the REQUIRE file happening to live in the package repository. That copy of the file actually doesn't get used at all when you install a tagged release of a package (only if you Pkg.clone or Pkg.checkout some branch). The METADATA requires file is the only copy of that information that gets used, and that's constantly mutable and can be edited by making a pull request to METADATA. You can change dependency bounds or platform information after-the-fact on existing tags, as long as you don't modify the sha that it points to.

**Comment by samoconnor** [comment]
Ok, I guess I've never seen a REQUIRE with platform rules.
This is the current REQUIRE for this package at version v0.5.8
```
julia 0.6
BinaryProvider
Compat 0.52.0
```

What can we add, so that when v0.5.9 is released, `Pkg.add` won't try to install v0.5.9 on platform other than `*-darwin*` or `*-linux*`?
And, what would we then change (via METADATA PR) so that `Pkg.add` will install v0.5.9 on windows and freebsd (assuming that testing finds that those platforms are working later on).

**Comment by ararslan** [comment]
The thing is, this package _should_ work fine on FreeBSD. We ship MbedTLS with Julia, so it's silly not to fall back to the one that's already available when we can't use binaries from BinaryBuilder.

**Comment by samoconnor** [comment]
@ararslan, of course I think we should make this work on FreeBSD. I'm sure @quinnj and @staticfloat can figure out the appropriate build script tweaks to achieve that.
However, I'm interested in knowing more about the general solution to the problem of how to release a package such that `Pkg.add` won't try to install it on platforms other than those that have been validated (and then what the followup METADATA PR should look like to enable a certain platform).
i.e. in the example above, what would we put in the REQUIRE file to limit v0.5.9 to `*-darwin*` and `*-linux*`?

**Comment by samoconnor** [comment]
(@ararslan sorry to have hijacked your Issue with this side-discussion)
